### PR TITLE
perf: implement recovery owned page batches and recycling

### LINKED ISSUE #1081: Task: Recovery-Owned Page Batches and Recycling
Planning document:
- `docs/tasks/000311-recovery-owned-page-batches-and-recycling.md`
- Type: task
- ID: 000311
- Title: Task: Recovery-Owned Page Batches and Recycling

Generated from document-first issue workflow.

## Summary

Replace owning hot-row redo materialization with recovery-specific decoding into
packed descriptors, then copy admitted payloads directly into reusable page-owned
batches. Workers consume the existing borrowed-value interfaces and return batch
storage through completion for coordinator-side recycling. Add coarse controls
for batch size and idle retained capacity, and measure recovery performance with
the default system allocator.

## Context

Source Backlogs:

- docs/backlogs/000202-recovery-payload-allocation-bottleneck-bulk-recycling.md

Issue Labels:

- type:perf
- priority:medium
- codex

[Task 000309](000309-pipelined-recovery-with-parallel-page-replay.md) established
page-local replay on finite pool jobs. Its allocator investigation found that
coordinator-allocated row vectors and outlined `MemVar` values are freed on
workers while decoding continues. Returning existing batches to the coordinator
for disposal improved the measured workload, but did not eliminate nested
allocations or establish the mechanism behind jemalloc's separate speedup.

[Task 000310](000310-borrowed-value-low-level-execution.md) is implemented and
provides `ValRef`, `RowValues`, `UpdateValues`, and shared validation, sizing, and
page-writing consumers. Writers copy input bytes into page-owned storage and
retain no borrowed input. This task builds on that prerequisite.

The design follows the local background draft
`docs/drafts/redesign-parallel-recovery.md`: coordinator decoding, independent
page-batch ownership, borrowed worker consumption, and completion-based recycling.
That draft is ignored by Git; the required decisions are recorded here so this
task is implementable from its worktree alone. There is no parent RFC. The scope
is one recovery transport replacement, with the existing scheduler, durable
formats, and completed borrowed-value interfaces as fixed boundaries.

Current implementation references:

- `recovery/stream.rs`: `TrxLogIterator`, whole-group decode before exposure,
  `RedoLogStream`, and shared startup/catalog-checkpoint consumers.
- `log/block_group.rs`, `log/redo.rs`, and `serde.rs`: transaction framing,
  owning row decoding, collection checks, and ordered-map semantics.
- `value.rs`: `Val::deser` copies byte values into `MemVar`.
- `recovery/mod.rs`: ordered classification and hot-page admission.
- `recovery/dispatch.rs`: `ReplayOp`, pending batches, accepted-job ownership,
  `BatchOutput`, and exclusive completion collection.
- `table/recover.rs`, `row/values.rs`, and `table/mod.rs`: batch mutation and
  existing borrowed consumers.

## Goals

- Remove per-row value/update vectors and per-value owning byte allocations from
  ordinary user-table hot-row decoding and transport.
- Preserve the current replay sequence, validation boundaries, diagnostics,
  checkpoint filtering, DDL fences, and accepted-job settlement.
- Reuse flat page-batch storage independently of retained page history.
- Keep the copy budget explicit: use the existing group buffer, record byte
  ranges during decoding, and copy admitted bytes directly into their batch.
- Add a batch byte target and a cap on actual idle vector capacities, with
  progress for oversized operations.
- Differentially test every value type and its boundary cases through the new
  decoder and packed representation, including exact floating-point bits.
- Demonstrate and explain performance and memory effects under the default
  allocator without making allocator replacement a deployment prerequisite.

## Non-Goals

- Changing public `Val`/`UpdateCol` ownership, persistent redo/page formats,
  foreground DML, undo ownership, or allocator policy in unrelated modules.
- Moving value decoding to workers, assigning fixed page partitions to threads,
  or changing the existing finite-job executor and page scheduling model.
- Shared group ownership across workers, zero-copy group transport, group-buffer
  pooling, or replacing coordinator routing maps with a new flat directory.
- Optimizing catalog/DDL value ownership or parallelizing cold deletes and hot
  index reconstruction.
- A total-recovery/RSS hard limit or streaming arbitrarily large transactions
  and groups; whole-group buffering and retained insertion history remain separate.
- Completing the broader allocator-mechanism investigation in backlog 000202.

### PR DESCRIPTION
**Pull request by jiangzhe**
Closes #1081

## Summary by CodeRabbit

* **New Features**
  * Recovery now supports configurable batch-size and recycled-memory limits.
  * Recovery batches flush based on operation count or memory usage.
  * Recovery memory can be recycled to reduce allocation overhead.
  * Recovery processing uses more efficient packed data handling.

* **Bug Fixes**
  * Invalid recovery settings and malformed log data are detected more reliably.
  * Recovery replay behavior is validated across varied data types and boundary cases.

* **Documentation**
  * Added documentation for recovery batching and redo-log format requirements.

### CODE REVIEW

**Comment by jiangzhe** [comment]
@codex review

File: doradb-storage/src/conf/consts.rs [hunk 1]
```diff
@@ -49,8 +49,13 @@ pub const DEFAULT_RECOVERY_DISABLE_DML_VALIDATION: bool = false;
...
 /// Default active replay pages per outstanding batch when sizing is automatic.
 pub const DEFAULT_RECOVERY_ACTIVE_PAGES_PER_TASK: usize = 4;
-/// Default maximum operations reserved in one replay batch.
+/// Default maximum operations admitted in one replay batch.
 pub const DEFAULT_RECOVERY_MAX_BATCH_OPS: usize = 256;
+/// Initial used-storage flush target for recovery page batches.
+pub const DEFAULT_RECOVERY_TARGET_BATCH_BYTES: usize = 256 * 1024;
+
+/// Maximum idle vector capacity retained by recovery batch recycling.
+pub const DEFAULT_RECOVERY_MAX_RECYCLED_BYTES: usize = 16 * 1024 * 1024;
```
[the remark this answers is not in the document]
Comment (jiangzhe) [reply doradb-storage/src/conf/consts.rs:58]: Fixed.

File: docs/backlogs/closed/000202-recovery-payload-allocation-bottleneck-bulk-recycling.md [hunk 2]
```diff
@@ -53,3 +53,11 @@ The baseline is commit `381921109c997c4ef108aafbab2080b3f86ea9c9`; pipeline meas
...
 
+## Close Reason
+
+- Type: implemented
+- Detail: Implemented via task 000311: recovery-owned page batches and bounded recycling address the identified allocation/free ownership bottleneck, with matched glibc and jemalloc measurements, production-path correctness tests, and failure/lifetime validation. Default-allocator replay improved 43.6% in the controlled 10-million-row comparison; the refined jemalloc implementation measured near parity with owning replay. The user accepted closure because the cause was identified and addressed; additional allocator comparisons and deeper mechanism isolation are not a current priority and are not claimed as completed. The separate component-level redo-format follow-up remains in backlog 000203.
+- Closed By: backlog close
+- Reference: [Task 000311](../../tasks/000311-recovery-owned-page-batches-and-recycling.md); [redo-format follow-up 000203](../000203-redo-value-component-and-contiguous-payloads.md).
```
[the remark this answers is not in the document]
Comment (jiangzhe) [reply docs/backlogs/closed/000202-recovery-payload-allocation-bottleneck-bulk-recycling.md:61]: old task docs are immutable. this is one rule. please add it.
